### PR TITLE
Atropos

### PR DESCRIPTION
**Pull request by semarpetrus**

### CODE REVIEW

**Comment by tanaes** [comment]
shouldn't need humann2 on this PR...

**Comment by antgonza** [comment]
Note that the tests are actually failing, see the log. Not sure if this is due the `date; ...; date` or the `travis_retry. Please try removing first the date ...

File: qp_shotgun/qc_trim/__init__.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from qiita_client import QiitaCommand
+
+from .qc_trim import qc_trim
+
+__all__ = ['qc_trim']
+
+# Define the kneaddata command
+req_params = {'input': ('artifact', ['per_sample_FASTQ'])}
+opt_params = {
+    'adapter': ['string', 'GATCGGAAGAGCACACGTCTGAACTCCAGTCAC'],  # 3' adapter
```
Comment (antgonza) [inline qp_shotgun/qc_trim/__init__.py]: Could you use better names (so they are easier to identify) for the users? For [example](https://github.com/qiita-spots/qp-deblur/blob/master/qp_deblur/__init__.py#L31)

File: qp_shotgun/qc_trim/qc_trim.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    reverse reads in some samples, for example after trimming and QC.
+    """
+    # we match filenames, samples, and run prefixes
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+    cmds = []
+
+    param_string = _format_qc_trim_params(parameters)
+
+    for run_prefix, sample, f_fp, r_fp in samples:
+        cmds.append('atropos trim --threads 4 %s -o %s -p %s -pe1 %s -pe2 %s'
```
Comment (antgonza) [inline qp_shotgun/qc_trim/qc_trim.py]: please make `threads` a parameter too, like [this](https://github.com/qiita-spots/qp-shotgun/blob/master/qp_shotgun/humann2/__init__.py#L44)

File: qp_shotgun/qc_trim/qc_trim.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        smd = partial(join, out_dir)
+        for suff in suffixes:
+            fname = smd(suff % rp)
+            if exists(fname):
+                files.append(fname)
+            else:
+                missing_files.append(fname)
+
+    if not files:
+        # KneadData did not create any files, which means that no sequence
```
Comment (antgonza) [inline qp_shotgun/qc_trim/qc_trim.py]: We don't have KneadData anymore ... is this outdated?

File: qp_shotgun/qc_trim/qc_trim.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        # was kept after quality control and filtering for host data
+        raise ValueError("No sequences left after running Atropos")
+
+    return [ArtifactInfo('QC_Trim files', 'per_sample_FASTQ', files)]
+
+
+def qc_trim(qclient, job_id, parameters, out_dir):
+    """Run kneaddata with the given parameters
```
Comment (antgonza) [inline qp_shotgun/qc_trim/qc_trim.py]: kneaddata?

**Review by antgonza — APPROVED** [review]
Some comments/requests.

File: qp_shotgun/qc_trim/qc_trim.py (hunk 3 above)
Comment (semarpetrus) [reply qp_shotgun/qc_trim/qc_trim.py]: No, we essentially copied the kneadData file and changed the functions within. I have not update the documentation yet.

File: qp_shotgun/qc_trim/qc_trim.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    samples have both a forward and a reverse sequence. However, the read
+    trimming step can sometimes eliminate all reverse reads, especially in low
+    coverage samples with poor overall reverse read quality.
+    """
+
+    # sort forward seqs
+    forward_seqs.sort()
+
+    # check that rev seqs are same len
+    if reverse_seqs:
```
Comment (tanaes) [inline qp_shotgun/qc_trim/qc_trim.py:62]: @antgonza is this a test that we should consider offloading to the fastq validator?

File: qp_shotgun/qc_trim/tests/test_qc_trim.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+               ('s2', 'SKD8.640184', './folder/s2_S011_L001_R1.fastq.gz',
+                None),
+               ('s3', 'SKB7.640196', './folder/s3_S013_L001_R1.fastq.gz',
+                None)]
+
+        obs = make_read_pairs_per_sample(fwd_fp, rev_fp, fp)
+
+        self.assertEqual(obs, exp)
+
+    def test_make_read_pairs_per_sample_match_fwd_rev_diffnum(self):
```
Comment (tanaes) [inline qp_shotgun/qc_trim/tests/test_qc_trim.py:116]: Is the name of this test swapped with the next one (line 134)?

**Review by tanaes — APPROVED** [review]
Looks awesome.

**Comment by tanaes** [comment]
This looks great! 

Any idea what this line in the Travis output means?

```
/home/travis/.travis/job_stages: line 169:  5258 Terminated              travis_jigger $! $timeout $cmd
```

File: qp_shotgun/qc_trim/qc_trim.py (hunk 5 above)
Comment (antgonza) [reply qp_shotgun/qc_trim/qc_trim.py:62]: I think that will be ideal ...

**Comment by antgonza** [comment]
That means that the tests ran out of time ... it needs to be fixed.

**Comment by tanaes** [comment]
That's kind of weird, no? It looks like the tests finished?

screenshot

**Comment by semarpetrus** [comment]
It is. The " Terminated Line" is because of travis_wait. From what I understood is that travis_wait essentially runs the code in the background and uses kill to stop it and that results in the terminated message.

**Comment by antgonza** [comment]
closing in favor of https://github.com/qiita-spots/qp-shotgun/pull/36

**Pull request closed without merging**